### PR TITLE
Discount duplication

### LINKED ISSUE #38688: Duplicate discounts
# Context
Back office users can duplicate discounts from the list.
All options are identical for a duplicate, except for:
- the name => “Copy of ...”
- the promo code, which must be regenerated if there was one
- the status, which must be disabled
- same priority (if there is one)

Users stays on the discount list when they create a duplicate (a pop-in is displayed when the duplicate is created displaying error or succes message)

# Mockup

Image

# Tests
- check name, status & promo code when a discount is duplicated
- edge case: duplicate of a discount with a name of 254 char 🤷

### PR DESCRIPTION
**Pull request by boherm**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add the discount duplication behaviour.<br>**Bonus:** fix the fact that combination types discount aren't correctly deleted when a discount is delete.
| Type?             | improvement
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | See #38688
| UI Tests          | https://github.com/boherm/ga.tests.ui.pr/actions/runs/19897045830
| Fixed issue or discussion?     | Fixes #38688
| Related PRs       | ~
| Sponsor company   | PrestaShop SA

### CODE REVIEW

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Actions`
    - [ ] `Duplicate discount`
- `Admin.Catalog.Notification`
    - [ ] `Remember to properly edit all information after duplicating.`

(Note: this is an automated message, but answering it will reach a real human)

File: src/Adapter/Discount/Repository/DiscountRepository.php [hunk 1]
```diff
@@ -49,7 +50,8 @@ class DiscountRepository extends AbstractObjectModelRepository
     public function __construct(
         protected readonly DiscountValidator $cartRuleValidator,
         protected readonly Connection $connection,
-        protected readonly string $dbPrefix
+        protected readonly string $dbPrefix,
+        protected readonly TranslatorInterface $translator
```
Comment (jolelievre) [inline src/Adapter/Discount/Repository/DiscountRepository.php]:
```suggestion
```
Not used

File: src/Adapter/Discount/Repository/DiscountRepository.php [hunk 2]
```diff
@@ -191,6 +193,54 @@ public function getCustomerGroups(DiscountId $discountId): array
...
     }
 
+    /**
+     * @return int[]
+     */
+    public function getCompatibleTypes(DiscountId $discountId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('*')
```
Comment (jolelievre) [inline src/Adapter/Discount/Repository/DiscountRepository.php]:
```suggestion
            ->select('crt.id_cart_rule_type')
```

File: src/Adapter/Discount/Repository/DiscountRepository.php [hunk 3]
```diff
@@ -191,6 +193,54 @@ public function getCustomerGroups(DiscountId $discountId): array
...
+    }
+
+    /**
+     * @return int[]
+     */
+    public function getGroups(DiscountId $discountId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('*')
```
Comment (jolelievre) [inline src/Adapter/Discount/Repository/DiscountRepository.php]:
```suggestion
            ->select('crg.id_group')
```

File: src/Adapter/Discount/Repository/DiscountRepository.php [hunk 4]
```diff
@@ -191,6 +193,54 @@ public function getCustomerGroups(DiscountId $discountId): array
...
+    }
+
+    /**
+     * @return int[]
+     */
+    public function getShops(DiscountId $discountId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('*')
```
Comment (jolelievre) [inline src/Adapter/Discount/Repository/DiscountRepository.php]:
```suggestion
            ->select('crs.id_shop')
```

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    }
+
+    /**
+     * Global process of duplicating a discount
+     */
+    public function duplicate(DiscountId $discountId): DiscountId
+    {
+        $oldDiscountId = $discountId->getValue();
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateBefore',
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]:
```suggestion
            'actionAdminDuplicateDiscountBefore',
```

I know this was copied from `ProductDuplicator`, but now I think the hook name for product was not accurate enough when I see this new example.

We could tell the two cases apart based on the parameters (id_discount instead of id_product), but it may not be ideal 🤔

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 6]
```diff
@@ -0,0 +1,331 @@
...
+            ['id_discount' => $oldDiscountId]
+        );
+        $newDiscountId = $this->duplicateDiscount($discountId);
+
+        $this->duplicateRelations($discountId, $newDiscountId);
+
+        $this->hookDispatcher->dispatchWithParameters(
+            'actionAdminDuplicateAfter',
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]: Same (open) question

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 7]
```diff
@@ -0,0 +1,331 @@
...
+     *
+     * @return array<int, string>
+     */
+    private function getNewProductName(array $oldDiscountLocalizedNames): array
+    {
+        $newLocalizedNames = [];
+        foreach ($oldDiscountLocalizedNames as $langId => $oldName) {
+            $langId = (int) $langId;
+            $namePattern = $this->translator->trans('copy of %s', [], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
+            $newName = sprintf($namePattern, $oldName);
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]:
```suggestion
            $newName = $this->translator->trans('copy of %s', [$oldName], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```
or maybe
```suggestion
            $newName = $this->translator->trans('copy of %s', ['%s' => $oldName], 'Admin.Catalog.Feature', Language::getLocaleById($langId));
```

But `sprintf` is redundant since `trans` already handles placeholders

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+                ->values([
+                    'id_cart_rule' => ':id_discount',
+                    'quantity' => ':quantity',
+                    'type' => ':type',
+                ])
+                ->setParameter('id_discount', $newDiscountId->getValue())
+                ->setParameter('quantity', $ruleGroup['quantity'])
+                ->setParameter('type', $ruleGroup['type'])
+                ->executeQuery();
+            $ruleGroupsIds[$ruleGroup['id_product_rule_group']] = $this->connection->lastInsertId();
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]: Hmm be careful, you should create a new array for dedicated new rule groups IDs

```suggestion
            $newRuleGroupsIds[$ruleGroup['id_product_rule_group']] = $this->connection->lastInsertId();
```

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+        }
+
+        // Then, we can copy the rules, linking them to the new group IDs
+        $rulesIds = [];
+        $qb = $this->connection->createQueryBuilder();
+        $rules = $qb
+            ->select('*')
+            ->from($this->dbPrefix . 'cart_rule_product_rule')
+            ->where('id_product_rule_group IN (:id_rule_groups)')
+            ->setParameter('id_rule_groups', array_keys($ruleGroupsIds), ArrayParameterType::INTEGER)
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]: If you use one array for old and new, at this moment you are fetching both the old and new rule groups
It may work out since the new groups are empty, but it's out of luck for now

```suggestion
            ->setParameter('id_rule_groups', array_keys($oldRuleGroupsIds), ArrayParameterType::INTEGER)
```

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+            ->fetchAllAssociative();
+
+        foreach ($rules as $rule) {
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_product_rule')
+                ->values([
+                    'id_product_rule_group' => ':id_rule_group',
+                    'type' => ':type',
+                ])
+                ->setParameter('id_rule_group', $ruleGroupsIds[$rule['id_product_rule_group']])
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]:
```suggestion
                ->setParameter('id_rule_group', $newRuleGroupsIds[$rule['id_product_rule_group']])
```

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 11]
```diff
@@ -0,0 +1,331 @@
...
+
+    /**
+     * Duplicates product rules associated to a discount
+     */
+    private function duplicateProductRules(DiscountId $oldDiscountId, DiscountId $newDiscountId): void
+    {
+        // First, we need to copy the rule groups (with their new IDs)
+        $ruleGroupsIds = [];
+        $qb = $this->connection->createQueryBuilder();
+        $ruleGroups = $qb
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php:247]: I'm thinking of something, maybe you could use `DiscountRepository::getProductRulesGroup` to get the whole data at once (and avoid uplicating similar code) It changes your logic/algorithm a bit but it should work

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 12]
```diff
@@ -0,0 +1,331 @@
...
+            $qb = $this->connection->createQueryBuilder();
+            $qb->insert($this->dbPrefix . 'cart_rule_product_rule')
+                ->values([
+                    'id_product_rule_group' => ':id_rule_group',
+                    'type' => ':type',
+                ])
+                ->setParameter('id_rule_group', $ruleGroupsIds[$rule['id_product_rule_group']])
+                ->setParameter('type', $rule['type'])
+                ->executeQuery();
+            $rulesIds[$rule['id_product_rule']] = $this->connection->lastInsertId();
```
Comment (jolelievre) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]: Same suggestion regarding the rule Ids, seems like you are mixing old and new values in the same array

File: tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php [hunk 13]
```diff
@@ -116,6 +117,26 @@ public function assertDiscount(string $discountReference, TableNode $tableNode):
         }
     }
 
+    /**
+     * @Then discount :discountReference shouldn't have the following properties:
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php]: I didn't understand at first but with the scenario context it makes, I think it's worth adding some explanation here like

```suggestion
     * We need to check that fields were modified but we don't know the new value, only the old one (example: code)
     * @Then discount :discountReference shouldn't have the following properties:
```

File: tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php [hunk 14]
```diff
@@ -406,134 +427,294 @@ public function assertDiscountIsNotFound(string $discountReference): void
         }
     }
 
-    protected function assertDiscountProperties(DiscountForEditing $discountForEditing, array $expectedData): void
+    protected function assertDiscountProperties(DiscountForEditing $discountForEditing, array $expectedData, bool $notOperation = false): void
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Context/Domain/Discount/DiscountFeatureContext.php]: Hmm, although now that I see all the implications here for one field, I'm thinking maybe we should only have a dedicated step for code (the only one checked this way), and it would be simpler?

File: tests/Integration/Behaviour/Features/Scenario/Discount/BO/duplicate_discount.feature [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    Then discount "discount_with_country_copy" should have the following properties:
+      | name[en-US]              | copy of Promotion |
+      | type                     | free_shipping     |
+      | minimum_product_quantity | 0                 |
+      | active                   | false             |
+    Then discount "discount_with_country_copy" should have the following properties:
+      | condition_type | items        |
+      | countries      | spain,france |
+
+  Scenario: Duplicate discount with supplier trigger
```
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Discount/BO/duplicate_discount.feature:133]: Ah this part is about product rules right? Maybe we can make it a bit more generic and add other conditions type of product rules (combinations, products, ...)

Maybe not exhaustively but a few different ones

File: src/Adapter/Discount/Update/DiscountDuplicator.php [hunk 16]
```diff
@@ -0,0 +1,331 @@
...
+    {
+        $characters = '123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ';
+        $charactersLength = strlen($characters);
+        $randomCode = '';
+        for ($i = 0; $i < $length; ++$i) {
+            $randomCode .= $characters[random_int(0, $charactersLength - 1)];
+        }
+
+        return $randomCode;
+    }
```
Comment (tleon) [inline src/Adapter/Discount/Update/DiscountDuplicator.php]: just a question not blocking. couldn't we re use the code generation from the BO part ?

**Review by tleon — DISMISSED** [review]

File: src/Adapter/Discount/Update/DiscountDuplicator.php (hunk 16 above)
Comment (boherm) [reply src/Adapter/Discount/Update/DiscountDuplicator.php]: We have this code generation only on JS part :/ But I will use RandomString instead with a new function :)

**Review by jolelievre — APPROVED** [review]
Thanks @boherm

**Review by tleon — APPROVED** [review]

**Review by paulnoelcholot — COMMENTED** [review]
Hello @boherm

I tested your PR and I found 2 bug : 

In the first discount, minimum purchase amount are in Tax excluded and shipping excluded. With the clone, they are included.

The discount code is TEST123 in the first discount, with the clone, the code is VL3OPJAW.

For me, the first behavior is a bug, but for the second, I'd like to have the opinion of @ibahloul-ps 

https://github.com/user-attachments/assets/867398c3-9328-4d1a-9175-c870b149dc11

**Comment by boherm** [comment]
Hi @paulnoelcholot!

- The selection of "Tax excluded" or "Shipping excluded" doesn't work properly on saving a discount. You can check that with your first created discount => you will see Tax & Shipping included instead of excluded. We need to create an issue about this, but it's not related to this PR. (here, we copy data.. so..)

- If you look at the issue of this PR, @ibahloul-ps explain that we should regenerate a code if a code is present in the discount.

So I don't know how to do next 🤷‍♂️

**Comment by paulnoelcholot** [comment]
Thanks @boherm ,

I created the issue: https://github.com/PrestaShop/PrestaShop/issues/40268

So it's good for me!

Thanks!

**Review by paulnoelcholot — APPROVED** [review]
